File: adapter/adapter.test.ts
import { KRW, USD } from "./currency";
import { ExchangeRateV1Adapter } from "./exchange-rate-v1-adapter";
import { ExchangeRateV1Repository } from "./repository/exchange-rate-v1-repository";
import { ExchangeRateV2Repository } from "./repository/exchange-rate-v2-repository";

test('Adapter Run', () => {
    const exchangeRateV1 = new ExchangeRateV1Repository();
    const exchangeRateV2 = new ExchangeRateV2Repository();
    const exchangeRateV1Adapter = new ExchangeRateV1Adapter(exchangeRateV1);

    const v1USDToKRW = exchangeRateV1.fromUSD(KRW);
    expect(v1USDToKRW).toEqual(1271.14);
    
    const v2USDToKRW = exchangeRateV2.getExchangeRate(USD, KRW);
    expect(v2USDToKRW).toEqual(1271.14);

    const adapterUSDToKRW = exchangeRateV1Adapter.getExchangeRate(USD, KRW);
    expect(v2USDToKRW).toEqual(1271.14);

    
    const v1KRWToUSD = exchangeRateV1.fromKRW(USD);
    expect(v1KRWToUSD).toEqual(0.00079);
    
    const v2KRWToUSD = exchangeRateV2.getExchangeRate(KRW, USD);
    expect(v2KRWToUSD).toEqual(0.00079);

    const adapterKRWToUSD = exchangeRateV1Adapter.getExchangeRate(KRW, USD);
    expect(adapterKRWToUSD).toEqual(0.00079);
});

File: adapter/adapter_test.go
package adapter

import (
	"github.com/stretchr/testify/assert"
	"testing"
)

func TestAdapter(t *testing.T) {
	t.Run("Test Adapter", func(t *testing.T) {
		exchangeRateV1 := NewExchangeRateV1Repository()
		exchangeRateV2 := NewExchangeRateV2Repository()
		exchangeRateV1Adapter := NewExchangeRateV1Adapter(exchangeRateV1)

		v1USDToKRW, err := exchangeRateV1.FromUSD(KRW)
		if err != nil {
			t.Error(err)
		}
		assert.Equal(t, 1271.14, v1USDToKRW)

		v2USDToKRW, err := exchangeRateV2.GetExchangeRate(USD, KRW)
		if err != nil {
			t.Error(err)
		}
		assert.Equal(t, 1271.14, v2USDToKRW)

		adapterUSDToKRW, err := exchangeRateV1Adapter.GetExchangeRate(USD, KRW)
		if err != nil {
			t.Error(err)
		}
		assert.Equal(t, 1271.14, adapterUSDToKRW)

		v1KRWToUSD, err := exchangeRateV1.FromKRW(USD)
		if err != nil {
			t.Error(err)
		}
		assert.Equal(t, 0.00079, v1KRWToUSD)

		v2KRWToUSD, err := exchangeRateV2.GetExchangeRate(KRW, USD)
		if err != nil {
			t.Error(err)
		}
		assert.Equal(t, 0.00079, v2KRWToUSD)

		adapterKRWToUSD, err := exchangeRateV1Adapter.GetExchangeRate(KRW, USD)
		if err != nil {
			t.Error(err)
		}
		assert.Equal(t, 0.00079, adapterKRWToUSD)
	})
}
